cranelift/interpreter/src/step.rs

```diff
@@ -0,0 +1,717 @@
+        Opcode::SwidenLow => unimplemented!("SwidenLow"),
+        Opcode::SwidenHigh => unimplemented!("SwidenHigh"),
+        Opcode::UwidenLow => unimplemented!("UwidenLow"),
+        Opcode::UwidenHigh => unimplemented!("UwidenHigh"),
+        Opcode::FcvtToUint => unimplemented!("FcvtToUint"),
+        Opcode::FcvtToUintSat => unimplemented!("FcvtToUintSat"),
+        Opcode::FcvtToSint => unimplemented!("FcvtToSint"),
+        Opcode::FcvtToSintSat => unimplemented!("FcvtToSintSat"),
+        Opcode::FcvtFromUint => unimplemented!("FcvtFromUint"),
+        Opcode::FcvtFromSint => unimplemented!("FcvtFromSint"),
+        Opcode::Isplit => unimplemented!("Isplit"),
+        Opcode::Iconcat => unimplemented!("Iconcat"),
+        Opcode::AtomicRmw => unimplemented!("AtomicRmw"),
+        Opcode::AtomicCas => unimplemented!("AtomicCas"),
+        Opcode::AtomicLoad => unimplemented!("AtomicLoad"),
+        Opcode::AtomicStore => unimplemented!("AtomicStore"),
+        Opcode::Fence => unimplemented!("Fence"),
+
+        // TODO: these instructions should be removed once the new backend makes these obsolete
+        // (see https://github.com/bytecodealliance/wasmtime/issues/1936); additionally, the
+        // "all-arch" feature for cranelift-codegen would become unnecessary for this crate.
+        Opcode::X86Udivmodx
+        | Opcode::X86Sdivmodx
+        | Opcode::X86Umulx
+        | Opcode::X86Smulx
+        | Opcode::X86Cvtt2si
+        | Opcode::X86Vcvtudq2ps
+        | Opcode::X86Fmin
+        | Opcode::X86Fmax
+        | Opcode::X86Push
+        | Opcode::X86Pop
+        | Opcode::X86Bsr
+        | Opcode::X86Bsf
+        | Opcode::X86Pshufd
+        | Opcode::X86Pshufb
+        | Opcode::X86Pblendw
+        | Opcode::X86Pextr
+        | Opcode::X86Pinsr
+        | Opcode::X86Insertps
+        | Opcode::X86Punpckh
+        | Opcode::X86Punpckl
+        | Opcode::X86Movsd
+        | Opcode::X86Movlhps
+        | Opcode::X86Psll
+        | Opcode::X86Psrl
+        | Opcode::X86Psra
+        | Opcode::X86Pmullq
+        | Opcode::X86Pmuludq
+        | Opcode::X86Ptest
+        | Opcode::X86Pmaxs
+        | Opcode::X86Pmaxu
+        | Opcode::X86Pmins
+        | Opcode::X86Pminu
+        | Opcode::X86Palignr
+        | Opcode::X86ElfTlsGetAddr
+        | Opcode::X86MachoTlsGetAddr => unimplemented!("x86 instruction: {}", inst.opcode()),
+    })
+}
+
+#[derive(Error, Debug)]
+pub enum StepError {
+    #[error("unable to retrieve value from SSA reference: {0}")]
+    UnknownValue(ValueRef),
+    #[error("unable to find the following function: {0}")]
+    UnknownFunction(FuncRef),
+    #[error("cannot step with these values")]
+    ValueError(#[from] ValueError),
+    #[error("failed to access memory")]
+    MemoryError(#[from] MemoryError),
+}
+
+/// Enumerate the ways in which the control flow can change based on a single step in a Cranelift
+/// interpreter.
+pub enum ControlFlow<'a, V> {
```

Comment:
Out of curiosity, how do your static analyses handle control flow joins and loops? I don't see anything in this interpreter to handle that.

Reply:
In the constant folding code, I added a `PhiValues` structure that scans through a function twice, first to capture all block parameters and secondly to map the SSA values that could be inputs to those block parameters (e.g. from a `jump`). I perform lookups into that mapping when I need to join. Perhaps there is something like `PhiValues` somewhere but I couldn't find it?

Reply:
We don't have anything generic AFAIK, but you could take a look at the LICM and GVN implementations and see if they have anything we can factor out.

I must admit that I'm more interested in the design of the trait though. Don't have anything off the top of my head, just think it will need something like a `join` or `meet` method depending which way we are modeling our lattices, as well as a "should I take both branches?" predicate.

Reply:
I'll upstream that constant folding after this PR is merged and we can iterate on the design.